katdal/applycal.py

```diff
@@ -307,103 +428,115 @@ def calc_correction_per_corrprod(dump, channels, cache, inputs,
     KeyError
         If input and/or cal product has no associated correction
     """
-    g_per_input = np.ones((len(inputs), len(channels)), dtype='complex64')
-    for product in cal_products:
-        for n, inp in enumerate(inputs):
-            sensor_name = 'Calibration/{}_correction_{}'.format(inp, product)
-            g_product = cache.get(sensor_name)[dump]
+    n_channels = channels.stop - channels.start
+    g_per_input = np.ones((len(params.inputs), n_channels), dtype='complex64')
+    for product in params.products.values():
+        for n in range(len(params.inputs)):
+            sensor = product[n]
```

Comment:
Yeah, I didn't want to mess around too much with CategoricalData. Doing it there would create a list entry per dump per categorical sensor, which could be quite a lot. It wasn't clear to me if that's only going to be for all sensors or only those that are extracted though, so it might not actually be that bad.

Reply:
I'll keep this in mind for any future refactoring.